src/main/java/org/junit/runners/ParentRunner.java

```diff
@@ -243,14 +244,7 @@ protected Statement withAfterClasses(Statement statement) {
      *         found, or the base statement
      */
     private Statement withClassRules(Statement statement) {
-        RuleContainer ruleContainer = new RuleContainer();
-        RuleContainerHelper.CURRENT.set(ruleContainer);
-        try {
-            ruleContainer.addTestRules(classRules());
-        } finally {
-            RuleContainerHelper.CURRENT.remove();
-        }
-        return ruleContainer.apply(null, getDescription(), null, statement);
+        return RunRules.applyAll(statement, classRules(), getDescription());
```

Comment:
We should not be applying rules in this method. Instead, we should return a `Statement` which applies the rules, so the rules are applied when the final statement is evaluated. That's why `RunRules` is a `Statement`.